typed-racket-lib/typed-racket/types/pairwise-intersect.rkt

```diff
@@ -0,0 +1,121 @@
+#lang racket
+
+;; This module provides a function pairwise-intersect for calculating a lower
+;; bound of two types with respect to the precision order (*not* the subtype
+;; order). Contract typechecking uses this function to determine the type of a
+;; contracted value, given the value's type and the contract's output type.
+
+(require "../utils/utils.rkt"
+         (rep type-rep prop-rep object-rep values-rep)
+         (types subtype)
+         (only-in (infer infer) intersect)
+         racket/match)
+
+(provide pairwise-intersect)
+
+(define (pairwise-intersect/arr s t)
+  (match* (s t)
+    [((Arrow: s-dom s-rest s-kws s-rng)
+      (Arrow: t-dom t-rest t-kws t-rng))
+     #:when (= (length s-kws) (length t-kws))
+     (make-Arrow (map pairwise-intersect s-dom t-dom)
+                 (and s-rest t-rest
+                      (pairwise-intersect s-rest t-rest))
+                 (map pairwise-intersect/kw s-kws t-kws)
+                 (pairwise-intersect s-rng t-rng))]
+    [(_ _)
+     (raise-arguments-error
+      'pairwise-intersect/arr
+      "pairwise-interescting unsupported for drest/different length kws"
+      "s" s
+      "t" t)]))
+
+(define (pairwise-intersect/kw s t)
+  (match* (s t)
+    [((Keyword: ks ts rs) (Keyword: kt tt rt))
+     #:when (and (eq? kt ks)
+                 (or rt (not rs)))
+     (make-Keyword ks (pairwise-intersect ts tt) rs)]
+    [(_ _)
+     (raise-arguments-error
+      'pairwise-intersect/kw
+      "keywords must match"
+      "s" s
+      "t" t)]))
+
+(define (pairwise-intersect/prop-set ps qs)
+  (match* (ps qs)
+    [((PropSet: p+ p-) (PropSet: q+ q-))
+     (make-PropSet (pairwise-intersect/prop p+ q+)
+                   (pairwise-intersect/prop p- q-))]))
+
+(define (pairwise-intersect/prop p1 p2)
+  (match* (p1 p2)
+    [(p p) p]
+    [(p (TrueProp:)) p]
+    [((TrueProp:) p) p]
+    [((TypeProp: o s) (TypeProp: o t))
+     (make-TypeProp o (pairwise-intersect s t))]
+    [((NotTypeProp: o s) (NotTypeProp: o t))
+     (make-NotTypeProp o (pairwise-intersect s t))]
+    [(_ _)
+     (raise-arguments-error
+      'pairwise-intersect/prop
```

Comment:
This is a pretty limited set of props we support intersecting... is raising an error here the right thing? (i.e. won't this error be raised often? For example if one of the props is `FalseProp`, or an `And` or an `Or`, etc)

Reply:
The errors probably aren't the right thing.

But here's some context: I initially implemented `pairwise-intersect` just to get a few examples working, and I thought raising an error might help me debug the cases I missed as I added more examples.